common/djangoapps/third_party_auth/decorators.py

```diff
@@ -30,3 +34,52 @@ def wrapped_view(request, *args, **kwargs):
         resp['X-Frame-Options'] = x_frame_option
         return resp
     return wraps(view_func, assigned=available_attrs(view_func))(wrapped_view)
+
+
+def tpa_hint_ends_existing_session(func):
+    """
+    Modifies a view function so that, if a tpa_hint URL parameter is received, the user is
+    already logged in, and the hinted SSO provider is so configured, the user is redirected
+    to a logout view and then back here, so that we can ensure that they're using the most
+    "fresh" information.
+    """
+
+    @wraps(func)
+    def inner(request, *args, **kwargs):
+        """
+        Check for a TPA hint in combination with a logged in user, and log the user out
+        if the hinted provider specifies that they should be.
+        """
+        sso_provider = None
+        provider_id = request.GET.get('tpa_hint')
+        logged_out = request.GET.get('logged_out') == 'yes'
```

Comment:
This isn't to check whether the user is logged out; it's used to check whether we passed the user to the logout view in the past; i.e., even if the user is logged in, we assume it's a result of them coming back here, getting kicked to the login screen because they _weren't_ logged in, getting logged in through the third-party auth provider, and _then_ coming back here, now toting a logged-in user with them.